### ISSUE TITLE
Dependabot does not realize dependency needs updating

### ISSUE THREAD

**Issue by ka3de** [dependabot/dependabot-core#3966]
Hi team, we were facing some difficulties having Dependabot create PRs for updated packages in our Golang repositories, so I decided to create a couple of test repos, one for the "test dependency" and one for the "test application".
Unfortunately Dependabot does not seem to realize that the dependency needs upgrade in the application code, and does not create the corresponding PR. Actually you can see on the Dependabot log that it tries to verify that dependency but does not detect an update available for it (see below for more details).

**Package ecosystem**
go mod

**Package manager version**
1.16

**Language version**
1.16

**Manifest location and content prior to update**
/go.mod and /go.sum
You can see the test repositories here, the [application repo](https://github.com/ka3de/dependabot-test), and the [dependency repo](https://github.com/ka3de/dependabot-test-lib).

**Updated dependency**
The dependency to be updated should be `github.com/ka3de/dependabot-test-lib`.

**What you expected to see, versus what you actually saw**
I would expect Dependabot open a PR to update the dependency.
Previously we had defined a "name filter" for dependencies in Dependabot manifest (you can see it in the repo as commented part), but even disabling it does not seem to make it work.

**Images of the diff or a link to the PR, issue or logs**
If I execute dependabot manually to check for updates the output is the following:
```
  proxy | time="2021-06-22T12:02:45Z" level=info msg="proxy starting" commit=d61dc19fc0dad31690c1bca6f08df34177a06472
  proxy | 2021/06/22 12:02:45 Listening (:1080)
updater | 2021-06-22T12:02:46.272409453 [164919043:main:WARN:src/devices/src/legacy/serial.rs:434] Detached the serial input due to peer close/error.
updater | time="2021-06-22T12:02:48Z" level=info msg="guest starting" commit=78f50f9d6e6c3d3763bdf7e548b0ab75c7f4a376
updater | time="2021-06-22T12:02:48Z" level=info msg="starting job..." fetcher_timeout=5m0s job_id=164919043 updater_timeout=45m0s updater_version=0.154.3-f5f6fab69777447c8ab83abe7dd542aee02fa72b
updater | I, [2021-06-22T12:02:51.820983 #8]  INFO -- sentry: ** [Raven] Raven 3.1.2 ready to catch errors
updater | warning: parser/current is loading parser/ruby27, which recognizes
updater | warning: 2.7.3-compliant syntax, but you are running 2.7.1.
updater | warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
updater | INFO <job_164919043> Starting job processing
  proxy | 2021/06/22 12:02:55 [002] GET https://api.github.com:443/repos/ka3de/dependabot-test
  proxy | 2021/06/22 12:02:55 [002] * authenticating github api request
  proxy | 2021/06/22 12:02:55 [002] 200 https://api.github.com:443/repos/ka3de/dependabot-test
  proxy | 2021/06/22 12:02:55 [004] GET https://api.github.com:443/repos/ka3de/dependabot-test/git/refs/heads/main
  proxy | 2021/06/22 12:02:55 [004] * authenticating github api request
  proxy | 2021/06/22 12:02:55 [004] 200 https://api.github.com:443/repos/ka3de/dependabot-test/git/refs/heads/main
  proxy | 2021/06/22 12:02:56 [007] GET https://github.com:443/ka3de/dependabot-test/info/refs?service=git-upload-pack
  proxy | 2021/06/22 12:02:56 [007] * authenticating git server request (host: github.com)
  proxy | 2021/06/22 12:02:56 [007] 200 https://github.com:443/ka3de/dependabot-test/info/refs?service=git-upload-pack
  proxy | 2021/06/22 12:02:56 [009] POST https://github.com:443/ka3de/dependabot-test/git-upload-pack
  proxy | 2021/06/22 12:02:56 [009] * authenticating git server request (host: github.com)
  proxy | 2021/06/22 12:02:56 [009] 200 https://github.com:443/ka3de/dependabot-test/git-upload-pack
  proxy | 2021/06/22 12:02:56 [011] POST https://github.com:443/ka3de/dependabot-test/git-upload-pack
  proxy | 2021/06/22 12:02:56 [011] * authenticating git server request (host: github.com)
  proxy | 2021/06/22 12:02:56 [011] 200 https://github.com:443/ka3de/dependabot-test/git-upload-pack
updater | INFO <job_164919043> Finished job processing
updater | time="2021-06-22T12:02:56Z" level=info msg="task complete" container_id=job-164919043-file-fetcher exit_code=0 job_id=164919043 step=fetcher
updater | I, [2021-06-22T12:02:57.687005 #9]  INFO -- sentry: ** [Raven] Raven 3.1.2 ready to catch errors
updater | warning: parser/current is loading parser/ruby27, which recognizes
updater | warning: 2.7.3-compliant syntax, but you are running 2.7.1.
updater | warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
updater | INFO <job_164919043> Starting job processing
updater | INFO <job_164919043> Starting update job for ka3de/dependabot-test
updater | INFO <job_164919043> Checking if github.com/ka3de/dependabot-test-lib 0.0.0-20210604145440-6449d46e0325 needs updating
updater | INFO <job_164919043> Latest version is 0.0.0-20210604145440-6449d46e0325
updater | INFO <job_164919043> No update needed for github.com/ka3de/dependabot-test-lib 0.0.0-20210604145440-6449d46e0325
updater | INFO <job_164919043> Finished job processing
updater | time="2021-06-22T12:02:59Z" level=info msg="task complete" container_id=job-164919043-updater exit_code=0 job_id=164919043 step=updater
```
So you can see that it tries to check for updates for the dependency, but fails to realize there are available.

Thank you.

**Comment by thepwagner** [comment]
> updater | INFO <job_164919043> Checking if github.com/ka3de/dependabot-test-lib 0.0.0-20210604145440-6449d46e0325 needs updating

Thanks for this excellent report, this line in particular stood out to me.

Dependabot skips checking go modules that have version strings matching Go [pseudo versions](https://golang.org/doc/modules/version-numbers#tmp_1), only tagged versions: https://github.com/dependabot/dependabot-core/blob/95913a028b20b88d98ef342427fcb37e9ebcce98/go_modules/lib/dependabot/go_modules/update_checker/latest_version_finder.rb#L49

For this test to work, you'll want to publish multiple semver-compatible versions of `dependabot-test-lib`, Dependabot should propose updates between those.

I think the decision to ignore pseuodoversions was made to reduce the number of PRs Dependabot creates: proposing an upgrade to each new commit would be a lot.

**Comment by ka3de** [comment]
Hi @thepwagner , thanks for the response and apologies for late reply.

I was able to get it working. So I already supposed that semver releases where taken into account for upgrade checking, but if I understand correctly, another requirement is that the dependency is "included" in the application with a semver compatible version, right? So for example, if my application has a dependency with this "go mod pseudo-version number", even if the dependency has a new release compatible with semver syntax, this won't be taken into account, is that right?

I finally got it working by having an "initial" semver version dependency in the go.mod file for the application (e.g.: v1.0.0) and then creating a new release from the dependency (v1.1.0). So I guess this is the case.
